[Bug]: With gp3 storage_type for aws_db_instance cannot increase storage with gp3, or change to gp3, with storage < 400GB

### Terraform Core Version

1.2.9

### AWS Provider Version

4.45.0

### Affected Resource(s)

aws_db_instance

### Expected Behavior

Should be able to make change to an RDS with gp3 storage (recently added) with < 400GB storage.

### Actual Behavior

If `iops` is not provided, the following error is thrown:

> api error InvalidParameterCombination: The storage type gp3 requires `iops` to be specified.

If the default 3000 IOPS for gp3 is provided in the `iops` attribute, but the storage size is smaller than 400GB, the following error is thrown:

> api error InvalidParameterCombination: You can't specify IOPS or storage throughput for engine postgres and a storage size less than 400.

Most interestingly, it's possible to create an RDS instance with gp3 and minimal storage (20GB) but increasing that (e.g. to 30GB) throws the above errors, and if the RDS already exists but with another storage type like gp2, it's not possible to upgrade to a gp3 in place.


### Relevant Error/Panic Output Snippet

_No response_

### Terraform Configuration Files
```
terraform {
  # Note: this should match the version in `.tool-versions`
  required_version = "~> 1.2"

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 4.9"
    }
    random = {
      source  = "hashicorp/random"
      version = "~> 3.3"
    }
  }
}

provider "aws" {
}


resource "aws_vpc" "vpc" {
  cidr_block           = "10.0.0.0/16" 
  enable_dns_support   = true
  enable_dns_hostnames = true
}

resource "aws_subnet" "public_subnet1" {
  vpc_id            = aws_vpc.vpc.id
  cidr_block        = "10.0.1.0/24"
  availability_zone = "ap-southeast-2a" 
}
resource "aws_subnet" "public_subnet2" {
  vpc_id            = aws_vpc.vpc.id
  cidr_block        = "10.0.2.0/24"
  availability_zone = "ap-southeast-2b" 
}

resource "aws_db_subnet_group" "rds" {
  name       = "rds"
  subnet_ids = [aws_subnet.public_subnet1.id, aws_subnet.public_subnet2.id]
}

resource "aws_db_instance" "rds" {
  identifier     = "mydb"
  engine         = "postgres"
  engine_version = "14.5" 
  instance_class = "db.m6g.large"
  multi_az       = true
  storage_type   = "gp3"

  db_name  = "mydb"
  username = "myuser"
  password = "abcdefgh" 

  db_subnet_group_name = aws_db_subnet_group.rds.name
  apply_immediately    = true

  allocated_storage     = 20
  max_allocated_storage = 1000
  storage_encrypted     = true
  skip_final_snapshot   = true
}
```


### Steps to Reproduce

Create the resources as described , then change allocated_storage to a larger number and apply, observe the failures as noted in the Actual Behavior section.

### Debug Output

_No response_

### Panic Output

_No response_

### Important Factoids

_No response_

### References

Relates https://github.com/hashicorp/terraform-provider-aws/issues/27702.
Relates https://github.com/hashicorp/terraform-provider-aws/pull/27670.

https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_Storage.html

### Would you like to implement a fix?

No

I can reproduce for MySQL with a slight modification of the existing acceptance test case:

```console
% make testacc TESTARGS='-run=TestAccRDSInstance_gp3MySQL' PKG=rds ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/rds/... -v -count 1 -parallel 3  -run=TestAccRDSInstance_gp3MySQL -timeout 180m
=== RUN   TestAccRDSInstance_gp3MySQL
=== PAUSE TestAccRDSInstance_gp3MySQL
=== CONT  TestAccRDSInstance_gp3MySQL
    instance_test.go:4761: Step 3/3 error: Error running apply: exit status 1
        
        Error: updating RDS DB Instance (tf-acc-test-5977403606277891875): operation error RDS: ModifyDBInstance, https response error StatusCode: 400, RequestID: 6086d274-e75b-445c-afc5-8f6b14bbba02, api error InvalidParameterCombination: You can't specify IOPS or storage throughput for engine mysql and a storage size less than 400.
        
          with aws_db_instance.test,
          on terraform_plugin_test.tf line 15, in resource "aws_db_instance" "test":
          15: resource "aws_db_instance" "test" {
        
--- FAIL: TestAccRDSInstance_gp3MySQL (566.04s)
FAIL
FAIL	github.com/hashicorp/terraform-provider-aws/internal/service/rds	571.102s
FAIL
make: *** [testacc] Error 1
```

--- a/internal/service/rds/instance.go
+++ b/internal/service/rds/instance.go
@@ -1888,8 +1888,12 @@ func dbInstancePopulateModify(input *rds_sdkv2.ModifyDBInstanceInput, d *schema.
 
 	if d.HasChanges("allocated_storage", "iops") {
 		needsModify = true
-		input.Iops = aws.Int32(int32(d.Get("iops").(int)))
 		input.AllocatedStorage = aws.Int32(int32(d.Get("allocated_storage").(int)))
+
+		// Send Iops if it has changed or not (StorageType == "gp3" and AllocatedStorage < threshold).
+		if d.HasChange("iops") || !isStorageTypeGP3BelowAllocatedStorageThreshold(d) {
+			input.Iops = aws.Int32(int32(d.Get("iops").(int)))
+		}
 	}
 
 	if d.HasChange("auto_minor_version_upgrade") {
@@ -2189,6 +2193,22 @@ func resourceInstanceImport(_ context.Context, d *schema.ResourceData, meta inte
 	d.Set("skip_final_snapshot", true)
 	d.Set("delete_automated_backups", true)
 	return []*schema.ResourceData{d}, nil
+}
+
+// See https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_Storage.html#gp3-storage.
+func isStorageTypeGP3BelowAllocatedStorageThreshold(d *schema.ResourceData) bool {
+	if storageType := d.Get("storage_type").(string); storageType != storageTypeGP3 {
+		return false
+	}
+
+	switch allocatedStorage, engine := d.Get("allocated_storage").(int), d.Get("engine").(string); engine {
+	case InstanceEngineMariaDB, InstanceEngineMySQL, InstanceEnginePostgres:
+		return allocatedStorage < 400
+	case InstanceEngineOracleEnterprise, InstanceEngineOracleEnterpriseCDB, InstanceEngineOracleStandard2, InstanceEngineOracleStandard2CDB:
+		return allocatedStorage < 200
+	}
+
+	return false
 }
 
 func dbSetResourceDataEngineVersionFromInstance(d *schema.ResourceData, c *rds.DBInstance) {

--- a/internal/service/rds/instance_test.go
+++ b/internal/service/rds/instance_test.go
@@ -4955,6 +4955,58 @@ func TestAccRDSInstance_storageThroughput(t *testing.T) {
 	})
 }
 
+func TestAccRDSInstance_storageTypePostgres(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping long-running test in short mode")
+	}
+
+	var v rds.DBInstance
+	resourceName := "aws_db_instance.test"
+	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ErrorCheck:               acctest.ErrorCheck(t, rds.EndpointsID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckInstanceDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccInstanceConfig_storageTypePostgres(rName, "gp2", 200),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckInstanceExists(resourceName, &v),
+					resource.TestCheckResourceAttr(resourceName, "allocated_storage", "200"),
+					resource.TestCheckResourceAttr(resourceName, "iops", "0"),
+					resource.TestCheckResourceAttr(resourceName, "storage_throughput", "0"),
+					resource.TestCheckResourceAttr(resourceName, "storage_type", "gp2"),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+				ImportStateVerifyIgnore: []string{
+					"apply_immediately",
+					"final_snapshot_identifier",
+					"password",
+					"skip_final_snapshot",
+					"delete_automated_backups",
+					"blue_green_update",
+				},
+			},
+			{
+				Config: testAccInstanceConfig_storageTypePostgres(rName, "gp3", 300),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckInstanceExists(resourceName, &v),
+					resource.TestCheckResourceAttr(resourceName, "allocated_storage", "300"),
+					resource.TestCheckResourceAttr(resourceName, "iops", "3000"),
+					resource.TestCheckResourceAttr(resourceName, "storage_throughput", "125"),
+					resource.TestCheckResourceAttr(resourceName, "storage_type", "gp3"),
+				),
+			},
+		},
+	})
+}
+
 func testAccCheckInstanceAutomatedBackups(s *terraform.State) error {
 	conn := acctest.Provider.Meta().(*conns.AWSClient).RDSConn
 
@@ -9979,4 +10031,29 @@ resource "aws_db_instance" "test" {
   storage_throughput = %[3]d
 }
 `, rName, iops, throughput))
+}
+
+func testAccInstanceConfig_storageTypePostgres(rName string, storageType string, allocatedStorage int) string {
+	return fmt.Sprintf(`
+data "aws_rds_engine_version" "default" {
+  engine = "postgres"
+}
+
+resource "aws_db_instance" "test" {
+  identifier           = %[1]q
+  engine               = data.aws_rds_engine_version.default.engine
+  engine_version       = data.aws_rds_engine_version.default.version
+  instance_class       = "db.m6g.large"
+  db_name              = "test"
+  password             = "avoid-plaintext-passwords"
+  username             = "tfacctest"
+  parameter_group_name = "default.${data.aws_rds_engine_version.default.parameter_group_family}"
+  skip_final_snapshot  = true
+
+  apply_immediately = true
+
+  storage_type      = %[2]q
+  allocated_storage = %[3]d
+}
+`, rName, storageType, allocatedStorage)
 }

--- a/internal/service/rds/instance_test.go
+++ b/internal/service/rds/instance_test.go
@@ -5146,6 +5146,14 @@ func testAccInstanceConfig_orderableClassMySQLGP3() string {
 	return testAccInstanceConfig_orderableClass("mysql", "general-public-license", "gp3", mySQLPreferredInstanceClasses)
 }
 
+func testAccInstanceConfig_orderableClassPostgres() string {
+	return testAccInstanceConfig_orderableClass("postgres", "postgresql-license", "standard", postgresPreferredInstanceClasses)
+}
+
+func testAccInstanceConfig_orderableClassPostgresGP3() string {
+	return testAccInstanceConfig_orderableClass("postgres", "postgresql-license", "gp3", postgresPreferredInstanceClasses)
+}
+
 func testAccInstanceConfig_orderableClassMariadb() string {
 	return testAccInstanceConfig_orderableClass("mariadb", "general-public-license", "standard", mariaDBPreferredInstanceClasses)
 }
@@ -6765,28 +6773,18 @@ resource "aws_db_instance" "test" {
 }
 
 func testAccInstanceConfig_EnabledCloudWatchLogsExports_postgreSQL(rName string) string {
-	return fmt.Sprintf(`
-data "aws_rds_engine_version" "default" {
-  engine = "postgres"
-}
-
-data "aws_rds_orderable_db_instance" "test" {
-  engine                     = data.aws_rds_engine_version.default.engine
-  engine_version             = data.aws_rds_engine_version.default.version
-  preferred_instance_classes = [%[1]s]
-}
-
+	return acctest.ConfigCompose(testAccInstanceConfig_orderableClassPostgres(), fmt.Sprintf(`
 resource "aws_db_instance" "test" {
   allocated_storage               = 10
   enabled_cloudwatch_logs_exports = ["postgresql", "upgrade"]
   engine                          = data.aws_rds_engine_version.default.engine
-  identifier                      = %[2]q
+  identifier                      = %[1]q
   instance_class                  = data.aws_rds_orderable_db_instance.test.instance_class
   password                        = "avoid-plaintext-passwords"
   username                        = "tfacctest"
   skip_final_snapshot             = true
 }
-`, postgresPreferredInstanceClasses, rName)
+`, rName))
 }
 
 func testAccInstanceConfig_maxAllocatedStorage(rName string, maxAllocatedStorage int) string {

--- a/internal/service/rds/instance_test.go
+++ b/internal/service/rds/instance_test.go
@@ -4853,6 +4853,58 @@ func TestAccRDSInstance_gp3Postgres(t *testing.T) {
 	})
 }
 
+func TestAccRDSInstance_gp3SQLServer(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping long-running test in short mode")
+	}
+
+	var v rds.DBInstance
+	resourceName := "aws_db_instance.test"
+	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ErrorCheck:               acctest.ErrorCheck(t, rds.EndpointsID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckInstanceDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccInstanceConfig_gp3(rName, testAccInstanceConfig_orderableClassSQLServerExGP3, 200),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckInstanceExists(resourceName, &v),
+					resource.TestCheckResourceAttr(resourceName, "allocated_storage", "200"),
+					resource.TestCheckResourceAttr(resourceName, "iops", "3000"),
+					resource.TestCheckResourceAttr(resourceName, "storage_throughput", "125"),
+					resource.TestCheckResourceAttr(resourceName, "storage_type", "gp3"),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+				ImportStateVerifyIgnore: []string{
+					"apply_immediately",
+					"final_snapshot_identifier",
+					"password",
+					"skip_final_snapshot",
+					"delete_automated_backups",
+					"blue_green_update",
+				},
+			},
+			{
+				Config: testAccInstanceConfig_gp3(rName, testAccInstanceConfig_orderableClassSQLServerExGP3, 300),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckInstanceExists(resourceName, &v),
+					resource.TestCheckResourceAttr(resourceName, "allocated_storage", "300"),
+					resource.TestCheckResourceAttr(resourceName, "iops", "3000"),
+					resource.TestCheckResourceAttr(resourceName, "storage_throughput", "125"),
+					resource.TestCheckResourceAttr(resourceName, "storage_type", "gp3"),
+				),
+			},
+		},
+	})
+}
+
 func TestAccRDSInstance_storageThroughput(t *testing.T) {
 	if testing.Short() {
 		t.Skip("skipping long-running test in short mode")
@@ -5223,6 +5275,10 @@ func testAccInstanceConfig_orderableClassMariadb() string {
 
 func testAccInstanceConfig_orderableClassSQLServerEx() string {
 	return testAccInstanceConfig_orderableClass("sqlserver-ex", "license-included", "standard", sqlServerPreferredInstanceClasses)
+}
+
+func testAccInstanceConfig_orderableClassSQLServerExGP3() string {
+	return testAccInstanceConfig_orderableClass("sqlserver-ex", "license-included", "gp3", sqlServerPreferredInstanceClasses)
 }
 
 func testAccInstanceConfig_orderableClassSQLServerSe() string {
@@ -9885,7 +9941,7 @@ resource "aws_db_instance" "test" {
   engine               = data.aws_rds_engine_version.default.engine
   engine_version       = data.aws_rds_engine_version.default.version
   instance_class       = data.aws_rds_orderable_db_instance.test.instance_class
-  db_name              = "test"
+  db_name              = data.aws_rds_engine_version.default.engine == "sqlserver-ex" ? null : "test"
   password             = "avoid-plaintext-passwords"
   username             = "tfacctest"
   parameter_group_name = "default.${data.aws_rds_engine_version.default.parameter_group_family}"

--- a/internal/service/rds/instance_test.go
+++ b/internal/service/rds/instance_test.go
@@ -4749,7 +4749,7 @@ func TestAccRDSInstance_BlueGreenDeployment_updateWithDeletionProtection(t *test
 	})
 }
 
-func TestAccRDSInstance_gp3(t *testing.T) {
+func TestAccRDSInstance_gp3MySQL(t *testing.T) {
 	if testing.Short() {
 		t.Skip("skipping long-running test in short mode")
 	}
@@ -4765,9 +4765,10 @@ func TestAccRDSInstance_gp3(t *testing.T) {
 		CheckDestroy:             testAccCheckInstanceDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccInstanceConfig_gp3(rName),
+				Config: testAccInstanceConfig_gp3(rName, testAccInstanceConfig_orderableClassMySQLGP3, 200),
 				Check: resource.ComposeAggregateTestCheckFunc(
 					testAccCheckInstanceExists(resourceName, &v),
+					resource.TestCheckResourceAttr(resourceName, "allocated_storage", "200"),
 					resource.TestCheckResourceAttr(resourceName, "iops", "3000"),
 					resource.TestCheckResourceAttr(resourceName, "storage_throughput", "125"),
 					resource.TestCheckResourceAttr(resourceName, "storage_type", "gp3"),
@@ -4785,6 +4786,68 @@ func TestAccRDSInstance_gp3(t *testing.T) {
 					"delete_automated_backups",
 					"blue_green_update",
 				},
+			},
+			{
+				Config: testAccInstanceConfig_gp3(rName, testAccInstanceConfig_orderableClassMySQLGP3, 300),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckInstanceExists(resourceName, &v),
+					resource.TestCheckResourceAttr(resourceName, "allocated_storage", "300"),
+					resource.TestCheckResourceAttr(resourceName, "iops", "3000"),
+					resource.TestCheckResourceAttr(resourceName, "storage_throughput", "125"),
+					resource.TestCheckResourceAttr(resourceName, "storage_type", "gp3"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccRDSInstance_gp3Postgres(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping long-running test in short mode")
+	}
+
+	var v rds.DBInstance
+	resourceName := "aws_db_instance.test"
+	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ErrorCheck:               acctest.ErrorCheck(t, rds.EndpointsID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckInstanceDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccInstanceConfig_gp3(rName, testAccInstanceConfig_orderableClassPostgresGP3, 200),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckInstanceExists(resourceName, &v),
+					resource.TestCheckResourceAttr(resourceName, "allocated_storage", "200"),
+					resource.TestCheckResourceAttr(resourceName, "iops", "3000"),
+					resource.TestCheckResourceAttr(resourceName, "storage_throughput", "125"),
+					resource.TestCheckResourceAttr(resourceName, "storage_type", "gp3"),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+				ImportStateVerifyIgnore: []string{
+					"apply_immediately",
+					"final_snapshot_identifier",
+					"password",
+					"skip_final_snapshot",
+					"delete_automated_backups",
+					"blue_green_update",
+				},
+			},
+			{
+				Config: testAccInstanceConfig_gp3(rName, testAccInstanceConfig_orderableClassPostgresGP3, 300),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckInstanceExists(resourceName, &v),
+					resource.TestCheckResourceAttr(resourceName, "allocated_storage", "300"),
+					resource.TestCheckResourceAttr(resourceName, "iops", "3000"),
+					resource.TestCheckResourceAttr(resourceName, "storage_throughput", "125"),
+					resource.TestCheckResourceAttr(resourceName, "storage_type", "gp3"),
+				),
 			},
 		},
 	})
@@ -9813,9 +9876,9 @@ data "aws_rds_orderable_db_instance" "updated" {
 `, rName, deletionProtection))
 }
 
-func testAccInstanceConfig_gp3(rName string) string {
+func testAccInstanceConfig_gp3(rName string, orderableClassConfig func() string, allocatedStorage int) string {
 	return acctest.ConfigCompose(
-		testAccInstanceConfig_orderableClassMySQLGP3(),
+		orderableClassConfig(),
 		fmt.Sprintf(`
 resource "aws_db_instance" "test" {
   identifier           = %[1]q
@@ -9831,9 +9894,9 @@ resource "aws_db_instance" "test" {
   apply_immediately = true
 
   storage_type      = data.aws_rds_orderable_db_instance.test.storage_type
-  allocated_storage = 200
+  allocated_storage = %[2]d
 }
-`, rName))
+`, rName, allocatedStorage))
 }
 
 func testAccInstanceConfig_storageThroughput(rName string, iops, throughput int) string {

